### PR TITLE
Serialize constant types

### PR DESCRIPTION
**Pull request by tbennun**
- [x] Serialize
- [x] Deserialize

### CODE REVIEW

File: dace/codegen/targets/framecode.py [hunk 1]
```diff
@@ -52,7 +51,8 @@ def generate_constants(self, sdfg: SDFG, callsite_stream: CodeIOStream):
                 callsite_stream.write(const_str, sdfg)
             else:
                 callsite_stream.write(
-                    "constexpr auto %s = %s;\n" % (cstname, str(cstval)), sdfg)
+                    "constexpr %s %s = %s;\n" % (csttype.dtype.ctype, cstname,
```
Comment (definelicht) [inline dace/codegen/targets/framecode.py]: This is great, now we can remove `generate_constants` from the FPGA codegen and just call this instead :-)

File: dace/sdfg.py [hunk 2]
```diff
@@ -220,13 +220,10 @@ def __init__(self,
...
             raise InvalidSDFGError('Invalid SDFG name "%s"' % name, self, None)
 
-        #if not isinstance(arg_types, collections.OrderedDict):
-        #    raise TypeError
-
-        #self._arg_types = arg_types  # OrderedDict(str, typeclass)
-        #self._constants = constants  # type: Dict[str, Any]
         self.arg_types = arg_types or collections.OrderedDict()
-        self.constants_prop = constants or {}
+        self.constants_prop = {}
```
Comment (definelicht) [inline dace/sdfg.py]: Why is it that we need a separate `constants` and `constants_prop`? Seems pretty ugly

File: dace/sdfg.py [hunk 3]
```diff
@@ -530,23 +528,49 @@ def constants(self):
...
+                return value
+            elif isinstance(dtype, dt.Scalar):
+                return dtype.dtype(value)
+            raise TypeError('Unsupported data type %s' % dtype)
+
+        result.update({k: cast(*v) for k, v in self.constants_prop.items()})
         return result
 
-    def add_constants(self, new_constants: Dict[str, Any]):
+    def add_constant(self, name: str, value: Any):
```
Comment (definelicht) [inline dace/sdfg.py]: I guess we cannot enforce that users don't directly update the constants dictionary, but we hope that they do this by convention?

File: dace/sdfg.py (hunk 2 above)
Comment (tbennun) [reply dace/sdfg.py]: Having only one structure would require many more changes in the code. Feel free to amend, no particular reason why it is this way.

File: dace/sdfg.py (hunk 3 above)
Comment (tbennun) [reply dace/sdfg.py]: We can hide away `add_constants` and make it an internal function, or just not exist at all. Your choice

File: dace/codegen/targets/framecode.py (hunk 1 above)
Comment (tbennun) [reply dace/codegen/targets/framecode.py]: :)

File: dace/sdfg.py (hunk 2 above)
Comment (definelicht) [reply dace/sdfg.py]: Ok, not worth it now

File: dace/sdfg.py (hunk 3 above)
Comment (definelicht) [reply dace/sdfg.py]: Why don't we let `add_constants` call `add_constant`?

Comment (tbennun) [reply dace/sdfg.py]: Sure. I can quickly write this up and let you know

Comment (tbennun) [reply dace/sdfg.py]: Fixed now

File: dace/sdfg.py [hunk 4]
```diff
@@ -220,13 +220,10 @@ def __init__(self,
...
-        #if not isinstance(arg_types, collections.OrderedDict):
-        #    raise TypeError
-
-        #self._arg_types = arg_types  # OrderedDict(str, typeclass)
-        #self._constants = constants  # type: Dict[str, Any]
         self.arg_types = arg_types or collections.OrderedDict()
-        self.constants_prop = constants or {}
+        self.constants_prop = {}
+        if constants is not None:
+            self.constants_prop.update(constants)
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/sdfg.py]: now it does

**Pull request merged**